modules/contact/src/AugmentedLagrangianContactProblem.C

```diff
@@ -0,0 +1,174 @@
+                                                             const Real abstol,
+                                                             const PetscInt nfuncs,
+                                                             const PetscInt /*max_funcs*/,
+                                                             const Real ref_resid,
+                                                             const Real /*div_threshold*/)
+{
+
+  Real my_max_funcs = std::numeric_limits<int>::max();
+  Real my_div_threshold = std::numeric_limits<Real>::max();
+
+  MooseNonlinearConvergenceReason reason =
+      ReferenceResidualProblem::checkNonlinearConvergence(msg,
+                                                          it,
+                                                          xnorm,
+                                                          snorm,
+                                                          fnorm,
+                                                          rtol,
+                                                          stol,
+                                                          abstol,
+                                                          nfuncs,
+                                                          my_max_funcs,
+                                                          ref_resid,
+                                                          my_div_threshold);
+
+  _console << "Augmented Lagrangian contact iteration " << _num_lagmul_iterations << "\n";
+
+  bool _augLM_repeat_step;
+
+  if (reason > 0)
+  {
+    if (_num_lagmul_iterations < _max_lagmul_iters)
+    {
+      NonlinearSystemBase & nonlinear_sys = getNonlinearSystemBase();
+      nonlinear_sys.update();
+
+      const ConstraintWarehouse & constraints = nonlinear_sys.getConstraintWarehouse();
+
+      std::map<std::pair<unsigned int, unsigned int>, PenetrationLocator *> * penetration_locators =
+          NULL;
+
+      bool displaced = false;
+      _augLM_repeat_step = false;
+      if (getDisplacedProblem() == NULL)
+      {
+        GeometricSearchData & geom_search_data = geomSearchData();
+        penetration_locators = &geom_search_data._penetration_locators;
+      }
+      else
+      {
+        GeometricSearchData & displaced_geom_search_data = getDisplacedProblem()->geomSearchData();
+        penetration_locators = &displaced_geom_search_data._penetration_locators;
+        displaced = true;
+      }
+
+      for (const auto & it : *penetration_locators)
+      {
+        PenetrationLocator & pen_loc = *(it.second);
+
+        BoundaryID slave_boundary = pen_loc._slave_boundary;
+
+        if (constraints.hasActiveNodeFaceConstraints(slave_boundary, displaced))
+        {
+          const auto & ncs = constraints.getActiveNodeFaceConstraints(slave_boundary, displaced);
+
+          for (const auto & nc : ncs)
+          {
+            if (!(std::dynamic_pointer_cast<MechanicalContactConstraint>(nc))
```

Comment:
You also need to handle the case where nc isn't a MechanicalContactConstraint object.